### PR TITLE
Internal simplifications to constructing datatypes

### PR DESCRIPTION
**Pull request by ajreynol**
In preparation for https://github.com/cvc5/cvc5/pull/8511.

This makes it so that unresolved sorts are automatically inferred when constructing datatypes at the NodeManager level.  This is in preparation for simplifying the API.

Changes:
(1) NodeManager is cleaned so that unresolved types are automatically inferred and are not a part of its public interface. Internal code generating datatypes is simplified as a result.
(2) Adds necessary utilities to TypeNode and cleans an unused flag
(3) The parser is cleaned to not track unresolved sorts in an ad-hoc manner.
(4) The API is patched to use the simpler interface.

### CODE REVIEW

File: src/expr/dtype.cpp [hunk 1]
```diff
@@ -143,6 +143,52 @@ size_t DType::cindexOfInternal(Node item)
...
+  // other parametric types.
+  std::unordered_set<TypeNode> csfTypes;
+  for (const std::shared_ptr<DTypeConstructor>& ctor : d_constructors)
+  {
+    for (size_t i = 0, nargs = ctor->getNumArgs(); i < nargs; i++)
+    {
+      Node sel = (*ctor)[i].d_selector;
+      if (sel.isNull())
+      {
+        // currently permit null selector for representing self selectors,
```
Comment (aniemetz) [inline src/expr/dtype.cpp:161]:
```suggestion
        // we currently permit null selector for representing self selectors,
```

File: src/expr/dtype.cpp [hunk 2]
```diff
@@ -143,6 +143,52 @@ size_t DType::cindexOfInternal(Node item)
   return item.getAttribute(DTypeConsIndexAttr());
 }
 
+void DType::getUnresolvedDatatypeTypes(std::set<TypeNode>& unresTypes) const
```
Comment (aniemetz) [inline src/expr/dtype.cpp:146]: Maybe we should coll this `collectUnresolvedDatatypeTypes`, `get` kinda implies that this would return something.

**Review by aniemetz — APPROVED** [review]
Very nice!! Two minor things, else LGTM.

File: src/expr/dtype.cpp (hunk 2 above)
Comment (ajreynol) [reply src/expr/dtype.cpp:146]: Sounds good.

**Pull request merged**